In [1]:
# Import libraries required for data loading and exploratory data analysis

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
# Load Geldium's delinquency dataset from the Excel file

df = pd.read_excel("../Data/Delinquency_prediction_dataset.xlsx")

In [3]:
df.head()

,Customer_ID,Age,Income,Credit_Score,Credit_Utilization,Missed_Payments,Delinquent_Account,Loan_Balance,Debt_to_Income_Ratio,Employment_Status,Account_Tenure,Credit_Card_Type,Location,Month_1,Month_2,Month_3,Month_4,Month_5,Month_6
0,CUST0001,56,165580.0,398.0,0.390502,3,0,16310.0,0.317396,EMP,18,Student,Los Angeles,Late,Late,Missed,Late,Missed,Late
1,CUST0002,69,100999.0,493.0,0.312444,6,1,17401.0,0.196093,Self-employed,0,Standard,Phoenix,Missed,Missed,Late,Missed,On-time,On-time
2,CUST0003,46,188416.0,500.0,0.359930,0,0,13761.0,0.301655,Self-employed,1,Platinum,Chicago,Missed,Late,Late,On-time,Missed,Late
3,CUST0004,32,101672.0,413.0,0.371400,3,0,88778.0,0.264794,Unemployed,15,Platinum,Phoenix,Late,Missed,Late,Missed,Late,Late
4,CUST0005,60,38524.0,487.0,0.234716,2,0,13316.0,0.510583,Self-employed,11,Standard,Phoenix,Missed,On-time,Missed,Late,Late,Late


In [4]:
df.shape

(500, 19)

## Dataset Overview

The dataset contains 500 customer records and 19 variables. Each row represents a customer, while the columns contain information related to customer demographics, financial characteristics, credit behaviour, employment status, account details, and monthly payment behaviour.

Understanding the dataset size and structure is important before performing further analysis because it provides an initial view of the available information and helps determine the appropriate EDA techniques.

## 1.1 Dataset Structure and Data Types

Before analysing patterns and risk factors, the data types of each variable need to be examined. This helps distinguish numerical variables from categorical variables and determines which analytical methods can be applied to each variable.

The dataset information will also help identify potential missing values that may require further investigation.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Customer_ID           500 non-null    object 
 1   Age                   500 non-null    int64  
 2   Income                461 non-null    float64
 3   Credit_Score          498 non-null    float64
 4   Credit_Utilization    500 non-null    float64
 5   Missed_Payments       500 non-null    int64  
 6   Delinquent_Account    500 non-null    int64  
 7   Loan_Balance          471 non-null    float64
 8   Debt_to_Income_Ratio  500 non-null    float64
 9   Employment_Status     500 non-null    object 
 10  Account_Tenure        500 non-null    int64  
 11  Credit_Card_Type      500 non-null    object 
 12  Location              500 non-null    object 
 13  Month_1               500 non-null    object 
 14  Month_2               500 non-null    object 
 15  Month_3               5

### Interpretation

The dataset contains 500 records and 19 variables. There are 5 numerical variables stored as `float64`, 4 numerical variables stored as `int64`, and 10 categorical variables stored as `object`.

Most variables contain complete data; however, three variables have missing values. `Income` has 461 non-null values, meaning 39 values are missing. `Credit_Score` has 498 non-null values, meaning 2 values are missing, and `Loan_Balance` has 471 non-null values, meaning 29 values are missing.

The presence of missing values in financial variables such as income, credit score, and loan balance needs to be investigated before further analysis because incomplete information could affect risk profiling and subsequent delinquency prediction.

## 1.2 Missing Data Analysis

Missing values can reduce the reliability of EDA and predictive modelling. Before deciding how to treat missing data, the number of missing values in each variable should be quantified so that the scale of the problem can be assessed.

In [6]:
# Count missing values in each column

missing_values = df.isnull().sum()

missing_values

Customer_ID              0
Age                      0
Income                  39
Credit_Score             2
Credit_Utilization       0
Missed_Payments          0
Delinquent_Account       0
Loan_Balance            29
Debt_to_Income_Ratio     0
Employment_Status        0
Account_Tenure           0
Credit_Card_Type         0
Location                 0
Month_1                  0
Month_2                  0
Month_3                  0
Month_4                  0
Month_5                  0
Month_6                  0
dtype: int64

### Interpretation

The missing-value analysis shows that only three variables contain missing data: `Income`, `Credit_Score`, and `Loan_Balance`.

`Income` has the highest number of missing values with 39 records, followed by `Loan_Balance` with 29 records and `Credit_Score` with 2 records. All remaining variables are complete.

Because these variables are relevant to financial risk assessment, the missing values should be investigated and appropriately treated before risk profiling and predictive modelling. The appropriate treatment will depend on the proportion of missing values, the distribution of each variable, and the potential impact of removing or imputing the records.

## 1.3 Missing Data Proportion

The number of missing values alone does not show how significant a data-quality issue is. Calculating the percentage of missing values relative to the total number of records helps assess the extent of missingness in each affected variable.

In [7]:
# Calculate the percentage of missing values in each column

missing_percentage = (df.isnull().sum() / len(df)) * 100

missing_percentage[missing_percentage > 0]

Income          7.8
Credit_Score    0.4
Loan_Balance    5.8
dtype: float64

## 1.4 Summary Statistics of Key Financial Variables

Before deciding how to handle missing values, the distributions of the affected numerical variables should be examined. Summary statistics such as the mean, median, minimum, maximum, and quartiles can help identify skewness and unusual values that may influence the choice of imputation method.

In [8]:
# Generate summary statistics for variables containing missing values

df[["Income", "Credit_Score", "Loan_Balance"]].describe()

,Income,Credit_Score,Loan_Balance
count,461.000000,498.000000,471.000000
mean,108379.893709,577.716867,48654.428875
std,53662.723741,168.881211,29395.537273
min,15404.000000,301.000000,612.000000
25%,62295.000000,418.250000,23716.500000
50%,107658.000000,586.000000,45776.000000
75%,155734.000000,727.250000,75546.500000
max,199943.000000,847.000000,99620.000000


### Interpretation

The summary statistics show that all three variables have different distributions and scales.

`Income` has a mean of approximately 108,380 and a median of 107,658. The mean and median are relatively close, suggesting that the distribution is not strongly skewed based on these summary statistics.

`Credit_Score` has a mean of approximately 577.7 and a median of 586. The values range from 301 to 847, and the mean and median are also relatively close.

`Loan_Balance` has a mean of approximately 48,654 and a median of 45,776. The mean is somewhat higher than the median, which may indicate some right-skewness in the distribution.

These results suggest that the missing values should not automatically be replaced using the mean. The median may provide a more robust imputation method, particularly for variables where extreme values or skewness may affect the mean.

## 2. Missing Data Treatment

Based on the missing-data proportions and summary statistics, median imputation will be used for `Income`, `Credit_Score`, and `Loan_Balance`.

Median imputation preserves all customer records and is less sensitive to extreme values than mean imputation. Synthetic data generation is not required at this stage because the amount of missing data is limited and can be addressed using a transparent statistical method.

In [9]:
# Fill missing values using the median of each respective variable

df["Income"] = df["Income"].fillna(df["Income"].median())
df["Credit_Score"] = df["Credit_Score"].fillna(df["Credit_Score"].median())
df["Loan_Balance"] = df["Loan_Balance"].fillna(df["Loan_Balance"].median())

In [10]:
# Verify that no missing values remain in the treated variables

df[["Income", "Credit_Score", "Loan_Balance"]].isnull().sum()

Income          0
Credit_Score    0
Loan_Balance    0
dtype: int64

### Verification

After applying median imputation, `Income`, `Credit_Score`, and `Loan_Balance` contain no remaining missing values. This confirms that the selected missing-data treatment was successfully applied and the dataset is ready for the next stage of data-quality assessment.

## 2.1 Duplicate Record Analysis

Duplicate records can distort exploratory analysis by giving some observations more weight than others. Checking for duplicates helps ensure that each customer record is represented appropriately in the dataset.

In [11]:
duplicate_count = df.duplicated().sum()
duplicate_count

np.int64(0)

### Interpretation

The duplicate check identified 0 completely duplicated records in the dataset. This indicates that no customer records are exact duplicates across all columns, so there is no need to remove duplicate rows at this stage.

The dataset therefore retains all 500 customer records for further analysis.

## 2.2 Categorical Data Consistency

Categorical variables can contain inconsistent values, such as unexpected categories, spelling differences, or formatting variations. Checking the unique values helps identify potential data-quality issues that could affect risk profiling and predictive modelling.

In [12]:
# Check the unique categories in Employment_Status

df["Employment_Status"].value_counts()

Employment_Status
Unemployed       93
retired          87
Employed         82
EMP              81
Self-employed    80
employed         77
Name: count, dtype: int64

### Interpretation

The `Employment_Status` variable contains inconsistent category formatting. `Employed`, `employed`, and `EMP` appear to represent the same employment category but are recorded using different capitalization and abbreviation formats. The `retired` category is also formatted differently from the other employment-status values.

These inconsistencies can cause the same underlying category to be treated as separate groups during analysis or predictive modelling. Therefore, the employment-status values should be standardized before analysing relationships between employment status and delinquency risk.

In [13]:
# Check the unique categories and their frequencies for credit card type

df["Credit_Card_Type"].value_counts()

Credit_Card_Type
Gold        118
Student     112
Business    108
Standard     86
Platinum     76
Name: count, dtype: int64

### Interpretation

The `Credit_Card_Type` variable contains five consistently formatted categories: Gold, Student, Business, Standard, and Platinum. No obvious duplicate categories caused by differences in capitalization, spelling, or abbreviations were identified.

Therefore, no standardization is required for this variable at this stage.

## 2.4 Checking Location Consistency

The `Location` variable will be reviewed for inconsistent category labels. This check helps ensure that the same location is not represented by multiple spellings or formats, which could affect customer-group analysis.

In [14]:
df["Location"].value_counts()

Location
Los Angeles    107
Phoenix        103
Chicago        103
Houston         95
New York        92
Name: count, dtype: int64

### Interpretation

The `Location` variable contains five location categories: Los Angeles, Phoenix, Chicago, Houston, and New York. The categories are consistently formatted, with no obvious differences in capitalization, spelling, or naming conventions.

Therefore, no standardization is required for the `Location` variable at this stage.

## 2.5 Checking Monthly Payment Status

The monthly payment-status variables will be reviewed to identify the categories used to represent customer payment behaviour. These variables are particularly important for delinquency risk profiling because they capture payment patterns over time.

In [15]:
# Check the categories and frequencies for the first monthly payment-status variable

df["Month_1"].value_counts()

Month_1
On-time    177
Missed     164
Late       159
Name: count, dtype: int64

### Interpretation

The `Month_1` variable contains three consistently formatted payment-status categories: `On-time`, `Missed`, and `Late`. There are no obvious spelling, capitalization, or formatting inconsistencies in these categories.

The distribution also shows that all three payment statuses occur frequently in the dataset, making monthly payment behaviour a potentially useful variable for further delinquency risk analysis. However, the relationship between these payment statuses and the delinquency outcome should be examined before drawing conclusions about their predictive importance.

## 2.6 Consistency Across Monthly Payment Status

The six monthly payment-status variables will be compared to confirm that they use a consistent set of categories. Consistency across these variables is important because they represent payment behaviour over multiple months and may later be used to identify delinquency patterns.

In [16]:
# Display the unique payment-status categories used in each month

for column in ["Month_1", "Month_2", "Month_3", "Month_4", "Month_5", "Month_6"]:
    print(f"\n{column}")
    print(df[column].value_counts())


Month_1
Month_1
On-time    177
Missed     164
Late       159
Name: count, dtype: int64

Month_2
Month_2
Late       173
Missed     167
On-time    160
Name: count, dtype: int64

Month_3
Month_3
Late       169
On-time    169
Missed     162
Name: count, dtype: int64

Month_4
Month_4
Late       181
Missed     160
On-time    159
Name: count, dtype: int64

Month_5
Month_5
Missed     187
On-time    162
Late       151
Name: count, dtype: int64

Month_6
Month_6
Late       172
Missed     168
On-time    160
Name: count, dtype: int64


### Interpretation

All six monthly payment-status variables use the same three categories: `On-time`, `Missed`, and `Late`. No additional or inconsistently formatted categories were identified across the six months.

The distribution varies slightly from month to month. For example, `Month_5` has the highest number of `Missed` payments (187), while `Month_4` has the highest number of `Late` payments (181). These variations indicate changes in payment behaviour across months and make the monthly payment-status variables useful candidates for further delinquency risk analysis.

At this stage, these observations describe the payment-status distribution only; their relationship with delinquency will be tested separately during risk profiling.

## 2.7 Standardizing Employment Status

The `Employment_Status` variable contains multiple labels representing employed customers, including `Employed`, `employed`, and `EMP`. These categories should be standardized to a single consistent label so that the same employment status is not treated as separate groups during analysis.

In [17]:
# Standardize employment-status labels to consistent category names

df["Employment_Status"] = df["Employment_Status"].replace({
    "employed": "Employed",
    "EMP": "Employed",
    "retired": "Retired"
})

In [19]:
df["Employment_Status"].value_counts()

Employment_Status
Employed         240
Unemployed        93
Retired           87
Self-employed     80
Name: count, dtype: int64

### Interpretation

After standardization, the `Employment_Status` variable contains four consistent categories: `Employed`, `Unemployed`, `Retired`, and `Self-employed`.

The previously separate labels `Employed`, `employed`, and `EMP` have been consolidated into a single `Employed` category, while `retired` has been standardized to `Retired`. This prevents the same underlying employment status from being treated as different categories during subsequent analysis.

The cleaned distribution shows that employed customers form the largest group in the dataset, with 240 records, followed by unemployed customers with 93, retired customers with 87, and self-employed customers with 80.

# 3. Risk Profiling and Delinquency Analysis

After addressing the major data-quality issues, the next step is to examine delinquency patterns and identify variables that may be associated with delinquency risk.

Understanding the distribution of the delinquency-related variable is important because it provides the baseline against which potential risk factors can be compared.

In [20]:
# Examine the distribution of the delinquency-related account indicator

df["Delinquent_Account"].value_counts()

Delinquent_Account
0    420
1     80
Name: count, dtype: int64

### Interpretation

The `Delinquent_Account` variable contains two categories: 0 and 1. There are 420 customers (84%) with a value of 0 and 80 customers (16%) with a value of 1.

For this analysis, `Delinquent_Account = 1` represents the delinquency class being examined, while `0` represents the non-delinquency class.

The distribution shows that the dataset contains substantially more non-delinquent customers than delinquent customers. This class imbalance is important to consider when evaluating relationships between customer characteristics and delinquency, as a variable may appear associated with delinquency simply because the two groups have different sizes.

In [21]:
# Calculate the percentage distribution of delinquency status

delinquency_distribution = df["Delinquent_Account"].value_counts(normalize=True) * 100

delinquency_distribution

Delinquent_Account
0    84.0
1    16.0
Name: proportion, dtype: float64

### Interpretation

The delinquency distribution shows that 84% of customers are classified as non-delinquent, while 16% are classified as delinquent.

This indicates that delinquent customers represent a smaller proportion of the dataset than non-delinquent customers. This class imbalance should be considered when analysing risk factors because comparisons should focus on delinquency rates within groups rather than relying only on the number of delinquent customers.

The 16% delinquency rate provides a baseline for comparing how delinquency varies across different customer characteristics.

## 3.1 Missed Payments and Delinquency

Missed payment history is an important potential risk indicator because repeated missed payments may be associated with delinquent accounts.

The relationship between `Missed_Payments` and `Delinquent_Account` will be examined to determine whether delinquency rates differ across customers with different numbers of missed payments.

In [22]:
# Calculate the number of customers and delinquency rate for each missed-payment level

missed_payment_risk = df.groupby("Missed_Payments")["Delinquent_Account"].agg(
    Customer_Count="count",
    Delinquency_Rate="mean"
)

missed_payment_risk["Delinquency_Rate"] *= 100

missed_payment_risk

,Customer_Count,Delinquency_Rate
Missed_Payments,,
0,77,18.181818
1,54,18.518519
2,79,16.455696
3,78,11.538462
4,83,14.457831
5,68,22.058824
6,61,11.475410


### Interpretation

The observed delinquency rate varies across different levels of missed payments rather than increasing consistently with the number of missed payments.

Customers with 5 missed payments have the highest observed delinquency rate at approximately 22.06%, while customers with 6 missed payments have a lower rate of approximately 11.48%. Customers with 1 missed payment have a delinquency rate of approximately 18.52%, compared with 18.18% for customers with no missed payments.

This indicates that the relationship between `Missed_Payments` and `Delinquent_Account` is not strictly monotonic in this dataset. The unexpected decrease at 6 missed payments should be investigated further, as it may reflect the relatively small group sizes, characteristics of the underlying customers, or other variables affecting delinquency. Therefore, `Missed_Payments` should be considered a potentially relevant risk indicator, but the observed results do not establish a simple linear relationship with delinquency.

## 3.2 Credit Utilization and Delinquency

Credit utilization represents the proportion of available credit being used by a customer. It is considered a potential risk indicator because higher utilization may be associated with greater financial pressure.

The relationship between `Credit_Utilization` and `Delinquent_Account` will be examined to determine whether customers with different utilization levels show different observed delinquency rates.

In [23]:
# Examine the distribution of credit utilization

df["Credit_Utilization"].describe()

count    500.000000
mean       0.491446
std        0.197103
min        0.050000
25%        0.356486
50%        0.485636
75%        0.634440
max        1.025843
Name: Credit_Utilization, dtype: float64

### Interpretation

The `Credit_Utilization` variable has 500 observations, with a mean of approximately 0.491 and a median of approximately 0.486. This indicates that the typical customer is using around half of their available credit.

The values range from approximately 0.05 to 1.026. Most customers fall between the 25th percentile of 0.356 and the 75th percentile of 0.634. The maximum value slightly exceeds 1.0, indicating that at least one customer has credit utilization above 100%.

The mean and median are relatively close, suggesting that the overall distribution is not strongly skewed. However, the value above 1.0 should be investigated because very high credit utilization may represent customers with unusually high credit usage or a potential data-quality issue.

Credit utilization will therefore be examined further by comparing utilization levels with `Delinquent_Account` to determine whether higher utilization is associated with different observed delinquency rates.

In [24]:
# Compare average credit utilization between delinquent and non-delinquent customers
credit_utilization_risk = df.groupby("Delinquent_Account")["Credit_Utilization"].agg(
    Customer_Count="count",
    Average_Utilization="mean",
    Median_Utilization="median"
)

credit_utilization_risk

,Customer_Count,Average_Utilization,Median_Utilization
Delinquent_Account,,,
0,420,0.488505,0.483751
1,80,0.506887,0.493021


### Interpretation

The comparison shows that delinquent customers have a slightly higher average credit utilization than non-delinquent customers.

- Non-delinquent customers: average utilization ≈ 0.489
- Delinquent customers: average utilization ≈ 0.507

The median utilization follows the same pattern:

- Non-delinquent customers: median ≈ 0.484
- Delinquent customers: median ≈ 0.493

This indicates a small positive association between credit utilization and observed delinquency in this dataset. However, the difference is relatively small, so credit utilization alone does not appear to strongly distinguish delinquent customers from non-delinquent customers.

Therefore, `Credit_Utilization` should be considered as one potential risk factor alongside other financial and behavioural variables rather than as a standalone indicator of delinquency.

## 3.3 Credit Score and Delinquency

Credit score is an important credit-risk variable that summarizes aspects of a customer's credit history and behaviour.

The relationship between `Credit_Score` and `Delinquent_Account` will be examined to determine whether delinquent and non-delinquent customers have different credit-score profiles.

In [25]:
# Compare credit scores between delinquent and non-delinquent customers
credit_score_risk = df.groupby("Delinquent_Account")["Credit_Score"].agg(
    Customer_Count="count",
    Average_Credit_Score="mean",
    Median_Credit_Score="median"
)

credit_score_risk

,Customer_Count,Average_Credit_Score,Median_Credit_Score
Delinquent_Account,,,
0,420,575.197619,584.0
1,80,591.150000,625.0


### Interpretation

The comparison shows that delinquent customers have a higher average and median credit score than non-delinquent customers in this dataset.

- Non-delinquent customers: average credit score ≈ 575.2, median = 584
- Delinquent customers: average credit score ≈ 591.2, median = 625

This is an unexpected pattern because the observed delinquent group does not have lower credit scores than the non-delinquent group. In this dataset, credit score therefore does not show the expected negative relationship with delinquency.

The difference should not be interpreted as evidence that higher credit scores cause delinquency. Other variables, interactions between risk factors, or characteristics of the synthetic dataset may contribute to this pattern.

Credit score should therefore be retained as a potential predictive variable, but its relationship with delinquency should be evaluated together with other risk factors rather than used as a standalone indicator.

## 3.4 Debt-to-Income Ratio and Delinquency

The Debt-to-Income Ratio (DTI) represents the relationship between a customer's debt obligations and income. A higher DTI can indicate a greater debt burden relative to income and may therefore be relevant when assessing delinquency risk.

The relationship between `Debt_to_Income_Ratio` and `Delinquent_Account` will be examined to determine whether delinquent and non-delinquent customers show different levels of debt burden.

In [26]:
# Compare debt-to-income ratios between delinquent and non-delinquent customers
dti_risk = df.groupby("Delinquent_Account")["Debt_to_Income_Ratio"].agg(
    Customer_Count="count",
    Average_DTI="mean",
    Median_DTI="median"
)

dti_risk

,Customer_Count,Average_DTI,Median_DTI
Delinquent_Account,,,
0,420,0.297445,0.301634
1,80,0.306301,0.303280


### Interpretation

The results show that delinquent customers have a slightly higher average and median Debt-to-Income Ratio than non-delinquent customers.

- Non-delinquent customers: average DTI ≈ 0.297, median ≈ 0.302
- Delinquent customers: average DTI ≈ 0.306, median ≈ 0.303

The difference is small, indicating that DTI shows only a weak positive association with observed delinquency in this dataset. Although delinquent customers have slightly higher debt-to-income ratios, the overlap between the two groups suggests that DTI alone is unlikely to be a strong discriminator of delinquency.

DTI should therefore be retained as a potential risk feature and evaluated together with other financial and behavioural variables.

## 3.5 Loan Balance and Delinquency

Loan balance represents the outstanding amount associated with a customer's loan. Higher outstanding balances may indicate greater financial exposure and could potentially be associated with delinquency risk.

The relationship between `Loan_Balance` and `Delinquent_Account` will be examined by comparing the average and median loan balances of delinquent and non-delinquent customers.

In [27]:
# Compare loan balances between delinquent and non-delinquent customers
loan_balance_risk = df.groupby("Delinquent_Account")["Loan_Balance"].agg(
    Customer_Count="count",
    Average_Loan_Balance="mean",
    Median_Loan_Balance="median"
)

loan_balance_risk

,Customer_Count,Average_Loan_Balance,Median_Loan_Balance
Delinquent_Account,,,
0,420,48555.135714,45776.0
1,80,48132.287500,45776.0


### Interpretation

The results show very similar loan balances for delinquent and non-delinquent customers.

- Non-delinquent customers: average loan balance ≈ 48,555, median = 45,776
- Delinquent customers: average loan balance ≈ 48,132, median = 45,776

The median loan balance is identical for both groups, while the average balance is slightly lower for delinquent customers. This indicates that `Loan_Balance` does not show a clear positive association with delinquency in this dataset.

Although loan balance remains relevant as a financial exposure variable, its standalone relationship with delinquency appears weak. It should therefore be considered alongside other customer characteristics and behavioural variables rather than treated as a primary delinquency indicator.

## 3.6 Monthly Payment Behaviour and Delinquency

The dataset contains six monthly payment-status variables representing customer payment behaviour over time.

Each month contains three possible statuses: `On-time`, `Late`, and `Missed`. Examining these variables against `Delinquent_Account` can help identify whether particular payment behaviours are associated with higher observed delinquency rates.

The analysis will compare delinquency rates across the payment-status categories for each month.

In [28]:
# Calculate delinquency rates for each monthly payment status
monthly_payment_risk = {}

for month in [f"Month_{i}" for i in range(1, 7)]:
    monthly_payment_risk[month] = (
        df.groupby(month)["Delinquent_Account"]
        .agg(Customer_Count="count", Delinquency_Rate="mean")
    )

    monthly_payment_risk[month]["Delinquency_Rate"] *= 100

monthly_payment_risk

{'Month_1':          Customer_Count  Delinquency_Rate
 Month_1                                  
 Late                159         11.320755
 Missed              164         16.463415
 On-time             177         19.774011,
 'Month_2':          Customer_Count  Delinquency_Rate
 Month_2                                  
 Late                173         15.028902
 Missed              167         15.568862
 On-time             160         17.500000,
 'Month_3':          Customer_Count  Delinquency_Rate
 Month_3                                  
 Late                169         14.792899
 Missed              162         17.283951
 On-time             169         15.976331,
 'Month_4':          Customer_Count  Delinquency_Rate
 Month_4                                  
 Late                181         19.337017
 Missed              160         16.875000
 On-time             159         11.320755,
 'Month_5':          Customer_Count  Delinquency_Rate
 Month_5                              

### Interpretation

The monthly payment analysis shows that delinquency rates vary across payment-status categories and months, but there is no consistent pattern in which `Missed` payments always correspond to the highest delinquency rate.

For example, in Month 1, customers with `On-time` payments have the highest observed delinquency rate (19.77%), compared with 16.46% for `Missed` and 11.32% for `Late`. In Month 4, the pattern changes, with `Late` payments having the highest delinquency rate (19.34%), while `On-time` payments have the lowest (11.32%).

Across the six months, the highest observed rate for a payment-status group is 19.77% for `On-time` customers in Month 1, while the lowest is 11.32% for `Late` customers in Month 1 and `On-time` customers in Month 4. These unexpected variations suggest that individual monthly payment status alone does not provide a simple or consistent explanation of delinquency in this dataset.

Monthly payment behaviour should therefore be retained as a potential predictive feature, but the observed patterns should be interpreted cautiously and investigated together with other variables. The unexpected relationship between payment status and delinquency is an important finding for the analytics team to investigate further.

## 3.7 Overall Missed-Payment Behaviour

The six monthly payment-status variables can be combined to identify the number of months in which each customer had a `Missed` payment.

This provides a simple measure of repeated missed-payment behaviour and allows us to examine whether customers with more frequent missed payments have different observed delinquency rates.

In [29]:
# Count the number of months with a missed payment for each customer
month_columns = [f"Month_{i}" for i in range(1, 7)]

df["Missed_Months"] = (df[month_columns] == "Missed").sum(axis=1)

# Compare delinquency rates across the number of missed-payment months
missed_months_risk = df.groupby("Missed_Months")["Delinquent_Account"].agg(
    Customer_Count="count",
    Delinquency_Rate="mean"
)

missed_months_risk["Delinquency_Rate"] *= 100

missed_months_risk

,Customer_Count,Delinquency_Rate
Missed_Months,,
0,40,20.000000
1,124,13.709677
2,177,17.514124
3,116,13.793103
4,34,17.647059
5,8,25.000000
6,1,0.000000


### Interpretation

The number of months with missed payments shows variation in observed delinquency rates, but the relationship is not consistently increasing as the number of missed months increases.

Customers with 0 missed months have an observed delinquency rate of 20.00%, while customers with 1, 2, 3, and 4 missed months have rates of approximately 13.71%, 17.51%, 13.79%, and 17.65%, respectively. The group with 5 missed months has the highest observed delinquency rate at 25.00%; however, this group contains only 8 customers, so the result should be interpreted cautiously. The 6-month group contains only one customer and has a 0% observed delinquency rate, making it unsuitable for drawing conclusions.

Overall, repeated missed-payment behaviour appears potentially relevant to delinquency prediction, but the relationship is not monotonic in this dataset. The small group sizes at the highest missed-month levels are an important limitation and should be considered when developing and evaluating a predictive model.

## 3.8 Key Risk Indicators and Unexpected Findings

Based on the analysis performed so far, the following variables show potentially relevant relationships with delinquency:

### Key Risk Indicators

- **Missed Payments:** Delinquency rates vary across different numbers of missed payments. However, the relationship is not consistently increasing, so missed payments should be considered together with other variables.

- **Credit Utilization:** Delinquent customers had a slightly higher average credit utilization (0.507) compared with non-delinquent customers (0.489), indicating a weak positive association.

- **Debt-to-Income Ratio:** Delinquent customers had a slightly higher average DTI (0.306) than non-delinquent customers (0.297), suggesting a weak positive association with delinquency.

- **Monthly Payment Behaviour:** Payment status varies in its relationship with delinquency across the six months. Certain months show higher delinquency rates for late or missed payments, but there is no consistent pattern across all months.

- **Repeated Missed Payments:** The number of months containing missed payments may provide additional behavioural information. However, delinquency rates are not monotonic across missed-payment counts, and groups with five or six missed months are very small.

### Unexpected Findings

- **Credit Score:** Delinquent customers had a higher average credit score (591.15) than non-delinquent customers (575.20), which is contrary to the relationship that might normally be expected from a credit-risk perspective. This should be investigated further rather than interpreted as a causal relationship.

- **Loan Balance:** Average loan balances were very similar between delinquent and non-delinquent customers (approximately 48,555 versus 48,132), suggesting little standalone separation between the two groups.

- **Payment Status:** In some months, customers classified as `On-time` had higher observed delinquency rates than customers classified as `Missed` or `Late`. This unexpected pattern indicates that individual payment-status variables may not independently explain delinquency in this dataset.

### Overall EDA Conclusion

The analysis indicates that delinquency is not explained by a single variable. Credit utilization and DTI show small differences between delinquent and non-delinquent customers, while payment behaviour and missed-payment measures show inconsistent relationships. Several unexpected patterns, particularly the higher credit scores among delinquent customers, should be investigated during predictive modeling. These findings support using multiple financial and behavioural variables together rather than relying on a single risk indicator.

# 4. EDA Summary

The dataset contains 500 customer records and 19 variables related to customer demographics, financial characteristics, credit behaviour, employment, account information, and monthly payment behaviour.

The data-quality assessment identified missing values in `Income`, `Credit_Score`, and `Loan_Balance`. Median imputation was applied to these variables, preserving all customer records. In addition, inconsistent employment-status labels were standardized, and no duplicate records were identified.

The delinquency analysis showed that 16% of customers were classified as delinquent. Several variables showed potentially relevant relationships with delinquency, including credit utilization, debt-to-income ratio, missed payments, and monthly payment behaviour. However, these relationships were generally weak or inconsistent.

Several unexpected findings were identified, including higher average credit scores among delinquent customers and similar loan balances between delinquent and non-delinquent customers. These findings demonstrate the importance of evaluating multiple variables together and validating observed patterns during predictive modeling.

The EDA provides a cleaned and better-understood dataset for the next stage of the analysis: developing and evaluating a delinquency prediction approach.